soundwire: SDCA: Generic interrupt support

mstrozek opened the pull request:
Generic interrupts using the kernel's regmap irqs

mstrozek committed:
ASoC: SDCA: Generic interrupt support

Add a library supporting usage of SDCA interrupts, using regmap irq
framework. The library adds functions for parsing ACPI for
interrupt-related information, configuring irq chip and requesting
individual irqs. Calling code (SDCA function code) is expected to also
substitute the library's base irq handler for its own, appropriate callback.

Signed-off-by: Maciej Strozek <mstrozek@opensource.cirrus.com>
Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

bardliao commented:
Thanks @mstrozek It looks quite decent to me. I will pause my SDCA interrupt work and check if Realtek codec drivers can also use the helpers. @shumingfan FYI

bardliao commented on include/sound/sdca_interrupts.h:
```diff
@@ -0,0 +1,79 @@
+
+/**
+ * struct sdca_interrupt_info - structure containing SDCA interrupt related information
+ * @parent: Pointer to the parent device.
```
Sorry, but, is it really the parent device? If so, do we really need both `struct device *dev;` and `void *parent` and why parent device is `void` type?

mstrozek replied:
you are correct, parent is wrong, changed to drv_data as per the below comment

bardliao commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,205 @@
+	}
+
+	interrupt_info->irq_chip = sdca_irq_chip;
+	interrupt_info->irq_chip.irq_drv_data = interrupt_info->parent;
```
Will `drv_data` be more readable than `parent`?

mstrozek replied:
yes, done

ranj063 commented on include/sound/sdca_interrupts.h:
```diff
@@ -0,0 +1,79 @@
+};
+
+/**
+ * struct sdca_interrupt - struct continaing information about an SDCA interrupt
```
typo containing

ranj063 commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,205 @@
+/*
+ * SDCA IRQ handler library
+ *
+ * Copyright (C) 2022-2023 Cirrus Logic, Inc. and
```
2022-2025?

charleskeepax replied:
Should really just be 2025 I think here, we havn't been working on this since 2022 :-)

ranj063 commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,205 @@
+{
+	struct sdca_interrupt *priv = data;
+
+	pr_err("%s IRQ handled\n", priv->name);
```
should this be removed?

mstrozek replied:
yes, removed now

ranj063 commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,205 @@
+	irq_handler_t thread_fn = irq->callback;
+	int ret;
+
+
```
extra line

ranj063 commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,205 @@
+
+
+	ret = irq_create_mapping(interrupt_info->irq_dom, irq->num);
+
```
no need for extra line here

ranj063 commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,205 @@
+static int sdca_process_controls_with_interrupts(struct sdca_interrupt_info *interrupt_info,
+					struct sdca_function_data *functions, int num_functions)
+{
+	int i, j, k;
```
can you please move this below the structs?

ranj063 commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,205 @@
+				control = &function->entities[j].controls[k];
+				if (control->interrupt_position > 0) {
+					if (total_controls >= SDCA_MAX_INTERRUPTS)
+						return -1;
```
-1 seems a bit odd. Would it be better to return total_controls here and do the check in the caller?

mstrozek replied:
Moved to the caller now

ranj063 commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,205 @@
+	int control_count;
+
+	control_count = sdca_process_controls_with_interrupts(interrupt_info, functions, num_functions);
+	if (control_count < 0) {
```
what if control_count is 0?

mstrozek replied:
should also fall under this condition, thanks for pointing it out

charleskeepax commented:
I think there might still be some issues with this patch, having some issues getting interrupts with it. Will update more once I have investigated a bit more.

charleskeepax commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,205 @@
+}
+
+#define IRQ_SDCA_OFFSET(reg_number) ((SDW_SCP_SDCA_INT##reg_number) - SDW_SCP_SDCA_INT1)
+#define IRQ_SDCA(number, reg_number) REGMAP_IRQ_REG(SDW_SCP_SDCA_INT_SDCA_##number, \
```
Should be just REGMAP_IRQ_REG(number, \

charleskeepax commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,205 @@
+	IRQ_SDCA(26, 4),
+	IRQ_SDCA(27, 4),
+	IRQ_SDCA(28, 4),
+	IRQ_SDCA(29, 4),
```
Missing IRQ_SDCA(30, 4),

charleskeepax commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,206 @@
+	.name = "sdca_irq",
+
+	.status_base = SDW_SCP_SDCA_INT1,
+	.mask_base = SDW_SCP_SDCA_INTMASK1,
```
SDCA IRQs are unmasked by writing a 1 so we should be using unmask_base. Also we need to write a 1 to ack the IRQ so we need to set ack_base = SDW_SCP_SDCA_INT1 as well.

charleskeepax commented on include/sound/sdca_interrupts.h:
```diff
@@ -0,0 +1,79 @@
+ * The MIPI SDCA specification is available for public downloads at
+ * https://www.mipi.org/mipi-sdca-v1-0-download
+ *
+ * Copyright(c) 2024 Intel Corporation
```
Should this file have Intel copyright or Cirrus? This is fine if you started from a file in one of Pierre's pull requests or something, but if you started from scratch should really be Cirrus copyright. Also typically we don't dual license source on the Cirrus side.

mstrozek replied:
I copy-pasted the license tag from another file and forgot to change it

charleskeepax commented on sound/soc/sdca/Makefile:
```diff
@@ -3,3 +3,7 @@
 
 obj-$(CONFIG_SND_SOC_SDCA)	+= snd-soc-sdca.o
+
+snd-soc-sdca-irq-handler-objs	:= sdca_interrupts.o
```
Should use -y rather than -objs

charleskeepax commented:
Ok I think the multiple IRQ thing was just caching on the regmap, so shouldn't affect this patch. Looks like I have basic IRQ functionality going on Phife (with my outstanding comments here). Will keep poking it all tomorrow.

plbossart commented on include/sound/sdca_interrupts.h:
```diff
@@ -0,0 +1,80 @@
+ * @irq_data: regmap irq chip data structure.
+ * @irq_dom: Pointer to the irq domain.
+ * @irqs_lock: Mutex to protect SDCA interrupts.
+ * @sdca_interrupt_register_list: Bitmask of SDCA interrupt registers.
```
all the kernel-doc is wrong, probably copy-pasted from my earlier work haha

plbossart commented on include/sound/sdca_interrupts.h:
```diff
@@ -0,0 +1,80 @@
+	struct regmap_irq_chip_data *irq_data;
+	struct irq_domain *irq_dom;
+
+	struct mutex irqs_lock; /* protects SDCA interrupts */
```
is this used and needed? I remember I had a good reason to add this lock but I can't remember which one....

mstrozek replied:
It is going to be needed for callbacks. Removed it now since nothing uses it directly, is it ok?

plbossart commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,207 @@
+	IRQ_SDCA(27, 4),
+	IRQ_SDCA(28, 4),
+	IRQ_SDCA(29, 4),
+	IRQ_SDCA(30, 4),
```
The register number seems overkill, it can be found directly from the interrupt number

mstrozek replied:
I thought it was easier to read with the reg number separate, but changed now to use only the irq number

plbossart commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,207 @@
+	case IRQF_TRIGGER_RISING:
+	case IRQF_TRIGGER_FALLING:
+		break;
+	case IRQ_TYPE_NONE:
```
I can't figure out how those flags would be used. Can we really support all of this?

mstrozek replied:
Probably not, you are right. Forced it now to a low level trigger, I guess this is closest in behaviour to soundwire's alert bit?

plbossart commented on include/sound/sdca.h:
```diff
@@ -11,6 +11,8 @@
 #include <linux/types.h>
 #include <linux/kconfig.h>
+#include <linux/regmap.h>
+#include <linux/interrupt.h>
```
Commit message needs more explanations and removed change-id

mstrozek replied:
Commit msg updated, is the wording ok?

bardliao commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,207 @@
+		dev_info(interrupt_info->dev, "No interrupts to act on\n");
+		return 0;
+	} else if (control_count >= SDCA_MAX_INTERRUPTS) {
+		dev_err(interrupt_info->dev, "Too many interrupts: %d\n", total_controls);
```
`total_controls` should be `control_count`

charleskeepax committed:
ASoC: SDCA: Add flag for unused IRQs

Zero is a valid SDCA IRQ interrupt position so add a special value to
indicate that the IRQ is not used.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

charleskeepax commented on include/sound/sdca_function.h:
```diff
@@ -1110,6 +1113,9 @@ struct sdca_function_data {
 	int num_clusters;
 
+	struct sdca_interrupt interrupt[SDCA_MAX_INTERRUPTS];
+	int interrupt_num;
```
I don't think this should live here, let the function driver store these, to be continued below...

charleskeepax commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,193 @@
+
+#define IRQ_SDCA_OFFSET(reg_number) ((SDW_SCP_SDCA_INT##reg_number) - SDW_SCP_SDCA_INT1)
+#define IRQ_SDCA(number) REGMAP_IRQ_REG(number, \
+						    IRQ_SDCA_OFFSET((number / 8) + 1), \
```
I don't think this will build, the string paste thing will happen before the preprocessor does the maths.

charleskeepax commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,193 @@
+	irq_handler_t thread_fn = irq->callback;
+	int ret;
+
+	ret = irq_create_mapping(interrupt_info->irq_dom, irq->num);
```
Better to use regmap_irq_get_virq, rather than bypassing the regmap IRQ API and using irq_create_mapping directly.

charleskeepax commented on sound/soc/sdca/sdca_interrupts.c:
```diff
@@ -0,0 +1,193 @@
+
+static int sdca_process_controls_with_interrupts(struct sdca_interrupt_info *interrupt_info,
+						 struct sdca_function_data *functions,
+						 int num_functions)
```
Continued... then this can become something that gets called from each function driver, so it no longer needs to manually loop over all the functions and is a little more in sync with the other APIs. I would also be inclined to split it into 2 functions one that returns an irq count for the function and that populates the structs. That is in keeping with all the asoc SDCA apis keeps everything consistent.

charleskeepax commented:
One last question I have, are we sure this should be a separate module, and should it be user selectable or selected by the the drivers using it? I am leaning towards keeping it as a separate module, but making it not user-selectable?

bardliao commented:
@mstrozek @charleskeepax Do you plan to submit the series upstream?

mstrozek commented:
@bardliao Yes, most likely next week after the merge window closes
